pkg/didcomm/protocol/route/service.go

```diff
@@ -296,6 +326,91 @@ func (s *Service) handleForward(msg *service.DIDCommMsg) error {
 	return s.outbound.Forward(forward.Msg, dest)
 }
 
+// Register registers the agent with a router with the on connectionID passed in. The agent is
+// registered with the router and retrieves the router endpoint and routing keys. This function
+// throws an error if the agent is already registered against a router. This function internally
+// works as a sync function and works in-conjunction with handleGrant function. ie, The request
+// did-comm message is sent to the router and it waits for the grant response from the router
+// till timeout.
+func (s *Service) Register(connectionID string) error {
+	// check if router is already registered
+	routerConnID, err := s.getRouterConnectionID()
+	if err != nil && !errors.Is(err, storage.ErrDataNotFound) {
+		return fmt.Errorf("fetch router connection id : %w", err)
+	}
+
+	if routerConnID != "" {
+		return errors.New("router is already registered")
+	}
```

Comment:
I'm not quite sure I understand this: if the data is not found, we return an error. If the data is found (?) we also return an error. 

saveRouterConnectionId() is only called from this method and is passed in the input connectionID as parameter.

I would have assumed that if getRouterConnectionID() returns ErrDataNotFound then the agent has not registered a route and therefore we can proceed.

Reply:
We are not returning error if the data is not found. The condition [!errors.Is(err, storage.ErrDataNotFound)](https://github.com/hyperledger/aries-framework-go/pull/1071/files#diff-7cc3df019152fe2c557826a312cc0a12R338) makes sure that we continue even if there is No data.

Reply:
D'Oh! My bad.